rs/sns/governance/src/governance.rs

```diff
@@ -3873,73 +3880,165 @@ impl Governance {
         // Safe to do with the validation step above
         let neuron_minimum_stake_e8s = self.neuron_minimum_stake_e8s_or_panic();
         let max_followees_per_function = self.max_followees_per_function_or_panic();
+        let max_number_of_principals_per_neuron =
+            self.max_number_of_principals_per_neuron_or_panic();
         let neuron_claimer_permissions = self.neuron_claimer_permissions_or_panic();
 
         let mut swap_neurons = vec![];
 
-        for neuron_parameter in &request.neuron_parameters {
-            match neuron_parameter.validate(neuron_minimum_stake_e8s, max_followees_per_function) {
-                Ok(_) => (),
-                Err(err) => {
-                    log!(ERROR, "Failed to claim Sale Neuron due to {:?}", err);
-                    swap_neurons.push(SwapNeuron::from_neuron_parameters(
-                        neuron_parameter,
-                        ClaimedSwapNeuronStatus::Invalid,
-                    ));
-                    continue;
+        // `request.neuron_parameters` is deprecated. For now,
+        // we will support both as long as only one is specified
+        #[allow(deprecated)] // TODO(NNS1-3198): Remove this once neuron_parameters is removed.
+        match (request.recipes, &request.neuron_parameters[..]) {
+            (Some(NeuronRecipes { recipes }), []) => {
+                for neuron_recipe in recipes {
+                    match neuron_recipe.validate(
+                        neuron_minimum_stake_e8s,
+                        max_followees_per_function,
+                        max_number_of_principals_per_neuron,
+                    ) {
+                        Ok(_) => (),
+                        Err(err) => {
+                            log!(ERROR, "Failed to claim Sale Neuron due to {:?}", err);
+                            swap_neurons.push(SwapNeuron::from_neuron_recipe(
+                                neuron_recipe,
+                                ClaimedSwapNeuronStatus::Invalid,
+                            ));
+                            continue;
+                        }
+                    }
+
+                    // Its safe to get all fields in NeuronRecipe because of the previous validation.
+                    let neuron_id = neuron_recipe.get_neuron_id_or_panic();
```

Comment:
NCR. Why not combine validation and field extraction?

Reply:
I think I would rather wait to do that until after we have a separation of API types and internal types. We could do that for this type in this PR, but this PR is already kind of large and complicated and I think keeping it as-is for now makes it easier to review (because you can easily see that it's doing pretty much the same thing for NeuronRecipe that it was doing for NeuronParameters

Reply:
> until after we have a separation of API types and internal types

I think we usually add such types opportunistically, so I thought maybe this might be that opportunity.

Reply:
> this PR is already kind of large and complicated

IMHO, part of the complexity comes from no doing things as per the usual convention. (Not to blame anyone, I know that the old code also didn't follow the current convention established in the NNS team).

Btw, I don't think the number of lines in the diff is a good metric for PR complexity. 

> you can easily see that it's doing pretty much the same thing for NeuronRecipe that it was doing for NeuronParameters

Exactly — my argument is that one cant easily see that it's doing the exactly right thing. I do see that the code is likely correct, but I cannot easily rule out all issues without having a clear view on what's going on e2e. Let's sync and discuss how this could be improved, or maybe you convince me that it's good enough to be merged as-is.

Reply:
I think if we apply [these changes](https://github.com/dfinity/ic/pull/679), that would be enough for now, as the function isn't yet panic-free anyway.